### PR TITLE
improve numbers parsing

### PR DESCRIPTION
**Pull request by SunyDays**
@KiChjang, need your review and comments
#10861

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10974)

### CODE REVIEW

File: components/style/attr.rs [hunk 1]
```diff
@@ -32,13 +33,11 @@ pub enum AttrValue {
...
-/// <https://html.spec.whatwg.org/multipage/#rules-for-parsing-non-negative-integers>
-fn do_parse_integer<T: Iterator<Item=char>>(input: T) -> Result<i64, ()> {
-    let mut input = input.skip_while(|c| {
-        HTML_SPACE_CHARACTERS.iter().any(|s| s == c)
-    }).peekable();
+fn do_parse_integer<T>(input: T) -> Result<i64, ()>
+    where T: Iterator<Item=char> + Clone {
+
+    let mut input = input.peekable();
+    input.take_while_ref(|c| HTML_SPACE_CHARACTERS.iter().any(|s| s == c));
```
Comment (KiChjang) [inline components/style/attr.rs]: I don't see the point of changing this.

Comment (SunyDays) [inline components/style/attr.rs]: after `skip_while` iterator `input` did not implement `Clone`which is necessary for `read_numbers`. I agree that it looks wired, maybe exists better solution?

File: tests/unit/style/attr.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@ use style::attr::{AttrValue, parse_length};
 use util::str::{DOMString, LengthOrPercentageOrAuto};
 
 #[test]
-fn test_parse_double() {
+fn test_from_double() {
```
Comment (KiChjang) [inline tests/unit/style/attr.rs]: There's really no need to change this.

File: components/util/str.rs [hunk 3]
```diff
@@ -159,72 +160,73 @@ fn is_ascii_digit(c: &char) -> bool {
...
-fn is_decimal_point(c: char) -> bool {
-    c == '.'
+fn is_decimal_point(c: &char) -> bool {
+    *c == '.'
 }
 
-fn is_exponent_char(c: char) -> bool {
-    match c {
+fn is_exponent_char(c: &char) -> bool {
+    match *c {
```
Comment (KiChjang) [inline components/util/str.rs]: I don't see the reason for the change here and above.

File: components/util/str.rs [hunk 4]
```diff
@@ -159,72 +160,73 @@ fn is_ascii_digit(c: &char) -> bool {
...
+        },
         Some(&'+') => {
             iter.next();
-            read_numbers(iter).0.map(|exp| exp.to_i32().unwrap_or(0))
-        }
-        Some(_) => read_numbers(iter).0.map(|exp| exp.to_i32().unwrap_or(0))
-    }
+            1
+        },
+        Some(_) | None => 1,
```
Comment (KiChjang) [inline components/util/str.rs]: You've changed the behaviour of `None` with your refactoring. Show me the spec that backs up this change.

File: components/util/str.rs [hunk 5]
```diff
@@ -159,72 +160,73 @@ fn is_ascii_digit(c: &char) -> bool {
...
-        });
-        (digits, accumulator.1 + 1)
+        })
     })
 }
 
-pub fn read_fraction<I: Iterator<Item=char>>(mut iter: Peekable<I>,
-                                             mut divisor: f64,
-                                             value: f64) -> (f64, usize) {
+pub fn read_fraction<I>(iter: &mut Peekable<I>, mut divisor: f64, value: &mut f64)
```
Comment (KiChjang) [inline components/util/str.rs]: Avoid mutable state as much as possible and return an f64.

File: components/style/attr.rs [hunk 6]
```diff
@@ -32,13 +33,11 @@ pub enum AttrValue {
...
 }
 
-/// Shared implementation to parse an integer according to
-/// <https://html.spec.whatwg.org/multipage/#rules-for-parsing-integers> or
-/// <https://html.spec.whatwg.org/multipage/#rules-for-parsing-non-negative-integers>
-fn do_parse_integer<T: Iterator<Item=char>>(input: T) -> Result<i64, ()> {
-    let mut input = input.skip_while(|c| {
-        HTML_SPACE_CHARACTERS.iter().any(|s| s == c)
-    }).peekable();
+fn do_parse_integer<T>(input: T) -> Result<i64, ()>
```
Comment (KiChjang) [inline components/style/attr.rs]: Perhaps this should take in a `&DOMString`, just to be consistent with `parse_double`.

File: components/style/attr.rs [hunk 7]
```diff
@@ -53,22 +52,26 @@ fn do_parse_integer<T: Iterator<Item=char>>(input: T) -> Result<i64, ()> {
...
-    let (value, _) = read_numbers(input);
+    let value = read_numbers(&mut input);
 
     value.and_then(|value| value.checked_mul(sign)).ok_or(())
 }
 
 /// Parse an integer according to
 /// <https://html.spec.whatwg.org/multipage/#rules-for-parsing-integers>.
-pub fn parse_integer<T: Iterator<Item=char>>(input: T) -> Result<i32, ()> {
+pub fn parse_integer<T>(input: T) -> Result<i32, ()>
```
Comment (KiChjang) [inline components/style/attr.rs]: Same for this function, take in a `&DOMString`.

File: components/style/attr.rs [hunk 8]
```diff
@@ -149,12 +147,11 @@ impl AttrValue {
...
-        let result = if result.is_infinite() {
-            default
-        } else {
-            result
-        };
+        let mut result = parse_double(&string).unwrap_or(default);
+        if result.is_infinite() {
+            result = default;
+        }
+
```
Comment (KiChjang) [inline components/style/attr.rs]: Avoid mutable state as much as possible.

**Comment by KiChjang** [comment]
@SunyDays Are you still planning to work on this?

**Comment by SunyDays** [comment]
@KiChjang, hi! yes, i will continue work on next week

**Comment by KiChjang** [comment]
@SunyDays It's been 2 weeks since your last comment, are you sure you are still going to continue working on this?

**Comment by KiChjang** [comment]
Closing due to inactivity.

**Pull request closed without merging**